pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/groupby/ObjectGroupByResultHolder.java

```diff
@@ -96,6 +129,83 @@ public void setValueForKey(int groupKey, double newValue) {
 
   @Override
   public void setValueForKey(int groupKey, Object newValue) {
-    _resultArray[groupKey] = newValue;
+    if (_storageMode == StorageMode.ARRAY_STORAGE) {
+      _resultArray[groupKey] = newValue;
+    } else {
+      _resultMap.put(groupKey, newValue);
+      _priorityQueue.add(new IntObjectPair(groupKey, newValue));
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Keys with 'lowest' values (as per the sort order) are trimmed away to reduce the
+   * size to _maxCapacity.
+   *
+   * @return Array of keys that were trimmed
+   */
+  @Override
+  public int[] trimResults() {
+    if (_storageMode == StorageMode.ARRAY_STORAGE) {
+      return EMPTY_ARRAY; // Still in array mode, trimming has not kicked in yet.
+    }
+
+    int size = _resultMap.size();
+    if (size <= _maxCapacity) {
+      return EMPTY_ARRAY;
+    }
+
+    int numKeysToRemove = size - _maxCapacity;
+    int[] removedGroupKeys = new int[numKeysToRemove];
+
+    for (int i = 0; i < numKeysToRemove; i++) {
+      IntObjectPair pair = _priorityQueue.poll();
+      int groupKey = pair.getIntValue();
+      _resultMap.remove(groupKey);
+      removedGroupKeys[i] = groupKey;
+    }
+    return removedGroupKeys;
+  }
+
+  /**
+   * Helper method to switch the storage from array mode to map mode.
+   */
+  private void switchToMapMode() {
+    _storageMode = StorageMode.MAP_STORAGE;
+    _resultMap = new Int2ObjectOpenHashMap(_resultArray.length);
+
+    _priorityQueue = getPriorityQueue(_resultArray.length, _minOrder);
+    for (int id = 0; id < _resultArray.length; id++) {
+      _resultMap.put(id, _resultArray[id]);
+      _priorityQueue.add(new IntObjectPair(id, (Comparable) _resultArray[id]));
+    }
+    _resultArray = null;
+  }
+
+  /**
+   * Returns a priority queue with the correct comparator as per
+   * the sort order specified.
+   *
+   * @return New priority queue with appropriate comparator based on sort order.
+   */
+  @SuppressWarnings("unchecked")
+  private PriorityQueue<IntObjectPair> getPriorityQueue(int initialSize, final boolean minOrder) {
+    return new PriorityQueue<>(initialSize, new Comparator<IntObjectPair>() {
```

Comment:
we probably have created these comparators elsewhere, can we move them to a ComparatorProvider/Util class